cannon/mipsevm/tests/testfuncs_test.go

```diff
@@ -0,0 +1,502 @@
+package tests
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/arch"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/exec"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded"
+	mttestutil "github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded/testutil"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/testutil"
+	preimage "github.com/ethereum-optimism/optimism/op-preimage"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+type operatorTestCase struct {
+	name      string
+	isImm     bool
+	rs        Word
+	rt        Word
+	imm       uint16
+	funct     uint32
+	opcode    uint32
+	expectRes Word
+}
+
+func testOperators(t *testing.T, cases []operatorTestCase, mips32Insn bool) {
+	versions := GetMipsVersionTestCases(t)
+	for _, v := range versions {
+		for i, tt := range cases {
+			// sign extend inputs for 64-bit compatibility
+			if mips32Insn {
+				tt.rs = randomizeUpperWord(signExtend64(tt.rs))
+				tt.rt = randomizeUpperWord(signExtend64(tt.rt))
+				tt.expectRes = signExtend64(tt.expectRes)
+			}
+
+			testName := fmt.Sprintf("%v (%v)", tt.name, v.Name)
+			t.Run(testName, func(t *testing.T) {
+				validator := testutil.NewEvmValidator(t, v.StateHashFn, v.Contracts)
+				goVm := v.VMFactory(nil, os.Stdout, os.Stderr, testutil.CreateLogger(), testutil.WithRandomization(int64(i)), testutil.WithPC(0), testutil.WithNextPC(4))
+				state := goVm.GetState()
+				var insn uint32
+				var baseReg uint32 = 17
+				var rtReg uint32
+				var rdReg uint32
+				if tt.isImm {
+					rtReg = 8
+					insn = tt.opcode<<26 | baseReg<<21 | rtReg<<16 | uint32(tt.imm)
+					state.GetRegistersRef()[rtReg] = tt.rt
+					state.GetRegistersRef()[baseReg] = tt.rs
+				} else {
+					rtReg = 18
+					rdReg = 8
+					insn = baseReg<<21 | rtReg<<16 | rdReg<<11 | tt.funct
+					state.GetRegistersRef()[baseReg] = tt.rs
+					state.GetRegistersRef()[rtReg] = tt.rt
+				}
+				testutil.StoreInstruction(state.GetMemory(), 0, insn)
+				step := state.GetStep()
+
+				// Setup expectations
+				expected := testutil.NewExpectedState(state)
+				expected.Step += 1
+				expected.PC = 4
+				expected.NextPC = 8
+				if tt.isImm {
+					expected.Registers[rtReg] = tt.expectRes
+				} else {
+					expected.Registers[rdReg] = tt.expectRes
+				}
+
+				stepWitness, err := goVm.Step(true)
+				require.NoError(t, err)
+
+				// Check expectations
+				expected.Validate(t, state)
+				validator.ValidateEVM(t, stepWitness, step, goVm)
+			})
+		}
+	}
+}
+
+type mulDivTestCase struct {
+	name         string
+	rs           Word
+	rt           Word
+	funct        uint32
+	opcode       uint32
+	expectHi     Word
+	expectLo     Word
+	expectRes    Word
+	rdReg        uint32
+	expectRevert string
+	errMsg       string
```

Comment:
(nit) Naming nits - the meaning of these is a little confusing given how they're handled in the test.  Maybe:
```suggestion
	panicMsg        string
	revertMsg       string
```